src/DeviceInterface/tests/initFocusApi.test.js

```diff
@@ -0,0 +1,308 @@
+        jest.restoreAllMocks();
+    });
+
+    it('should initialize and return the API methods', () => {
+        expect(focusApi).toBeDefined();
+        expect(typeof focusApi.setAutocompleteOnIdentityField).toBe('function');
+        expect(typeof focusApi.cleanup).toBe('function');
+    });
+
+    it('should register focus event listener on initialization', () => {
+        // We need to spy on window.addEventListener
+        const addEventSpy = jest.spyOn(window, 'addEventListener');
+
+        // Initialize the focus API
+        initFocusApi(forms, settings, true, attachKeyboardCallback);
+
+        // Check that addEventListener was called with expected args
+        expect(addEventSpy).toHaveBeenCalledWith('focus', expect.any(Function), true);
+
+        // Clean up spy
+        addEventSpy.mockRestore();
+    });
+
+    it('should handle focus events and call attachKeyboard when input_focus_api is enabled', () => {
+        // Mock pierce shadow tree to return our input element
+        mockedUtils.pierceShadowTree.mockReturnValue(mockInputElement);
+
+        // Create and dispatch a focus event
+        const focusEvent = new FocusEvent('focus');
+        window.dispatchEvent(focusEvent);
+
+        // Check that attachKeyboardCallback was called with correct arguments
+        expect(attachKeyboardCallback).toHaveBeenCalledWith({
+            form: mockForm,
+            element: mockInputElement,
+        });
+    });
+
+    it('should not call attachKeyboard when input_focus_api is disabled', () => {
+        // Disable the feature toggle
+        settings.featureToggles.input_focus_api = false;
+
+        // Reinitialize the API
+        focusApi.cleanup();
+        focusApi = initFocusApi(forms, settings, true, attachKeyboardCallback);
+
+        // Mock pierce shadow tree to return our input element
+        mockedUtils.pierceShadowTree.mockReturnValue(mockInputElement);
+
+        // Create and dispatch a focus event
+        const focusEvent = new FocusEvent('focus');
+        window.dispatchEvent(focusEvent);
+
+        // Check that attachKeyboardCallback was not called
+        expect(attachKeyboardCallback).not.toHaveBeenCalled();
+    });
+
+    it('should set autocomplete attribute when autocomplete_attribute_support is enabled', () => {
+        // Create an actual input element instead of mocking
+        const inputElement = document.createElement('input');
+
+        // Mock the behavior of all the dependencies
+        mockedUtils.pierceShadowTree.mockReturnValue(inputElement);
+        mockedMatching.getInputType.mockReturnValue('identities.emailAddress');
+        mockedMatching.getMainTypeFromType.mockReturnValue('identities');
+        mockedMatching.getSubtypeFromType.mockReturnValue('emailAddress');
+
+        // Spy on the methods we want to verify
+        const setAttributeSpy = jest.spyOn(inputElement, 'setAttribute');
+        const addEventListenerSpy = jest.spyOn(inputElement, 'addEventListener');
```

Comment:
Why mocking the whole world instead of checking that the right attributes are set on `focus` and then removed on `blur`? You can use the `attachAndReturnGenericForm` in _test-utils.js_ to make things simpler. You can pass a specific form with whatever elements you need. When you do that, you don't need to mock our functions like `pierceShadowTree`, `getInputType`. They will return the correct stuff because they'll read from the DOM. Unless I'm missing something. 🤔